### PR TITLE
wp3: next-run prep — diverse run config, holdout validator, mixture manifest prototype

### PR DESCRIPTION
**Pull request by josharmour**
## What (MORNING PRIORITY items 3+4 — config + prototype + validation only)

No training was run. No live-run file was touched (magezero `configs/run.yml` and `configs/run_diverse.yml` unmodified; deploy of the new config is a deliberate one-file copy, documented in the config header and PROGRESS file).

### Item 3 — diversity config for the NEXT curriculum run

`tools/training/wp3/next_run_diverse.yml` (tracked mtgacoach-side copy of the magezero `run_diverse.yml` draft, finalized):
- #438 greedy-optimal five opponents, exact `.dck` basenames verified against `deck_inventory.json`: `Oathbreaker_UR`, `GBLegends`, `EVG_Elves`, `EVG_Goblins`, `Mind(MindvsMight)` — plus the 5 retained Standard-Mono baselines. Card pool 83 -> 235 distinct (2.8x, from #438).
- `games_per_gen: 550` (draft had the pre-bump 200), `generations: 6`.
- Deploy (between runs only, never mid-run): `cp tools/training/wp3/next_run_diverse.yml /home/joshu/repos/magezero/configs/run_diverse.yml`

### Holdout contract — keeps the unseen-deck gate slice honest forever

`tools/training/wp3/validate_run_config.py` FAILS (exit 2) if any run config lists a permanent holdout deck (**HighNoonControl, BGRoots, BWBats**) as primary or opponent. Fails CLOSED on unreadable/unparseable/ambiguous configs. Normalized matching (case, whitespace, `.dck`). Wired as `tests/test_validate_run_config.py` (20 tests: contract pinned, every holdout tested both roles, all fail-closed branches, tracked config asserted clean + at 550x6 with the #438 five).

Measured against the live magezero configs dir:
```
[ok] /home/joshu/repos/magezero/configs/run.yml
[ok] /home/joshu/repos/magezero/configs/run_diverse.yml
OK: 2 config(s) clean of holdout decks (BGRoots, BWBats, HighNoonControl).
```

### Item 4 — breadth-vs-depth mixture manifest PROTOTYPE (owner picks the weight)

`tools/training/wp3/build_mixture_manifest.py --magezero-corpus ... --human-corpus ... --human-weight W --out manifest.json`
- MANIFEST ONLY — refuses to write training data without `--emit` (which also refuses overwrite and supports a `--min-production-share` fail-closed floor).
- Production-shape = the repo `Legal:` heuristic from `build_bridge_dataset.py` (`"Legal: (pick by number)"` in user, no `"Candidate:"`).
- Distinct-card coverage validated against `~/.arenamcp/mtgjson/name_index.json` (34,633 names + DFC-face aliases); unresolvable tokens counted + top-20 reported, never silently included.
- Trivial-policy floors per source (same reflex families as `gate_play_decisions.py` G6); `meta.gold_equivalent_picks` used where present (988 bridge records), single gold pick otherwise (lower bound, stated in manifest).
- 16 fixture-only unit tests.

**Human/Phase-0 corpora located on disk** (`tools/training/data/`, gitignored; all share `{system, user, response, source, attribution, kind, meta}`): `play_decisions.jsonl` (223 MB, 17lands, Candidate-shape), `play_decisions_nontrivial.jsonl` (13,671), `play_decisions_cast.jsonl`, `play_decisions_bridge.jsonl` (988, 100% production, real MTGA menus), `play_decisions_mixed.jsonl` (14,659, 6.74% production), `stage0_prompts_*` / `stage0_turnaction_*` / `stage0v2_*` (~0.8 GB, pre-WP3 prompt shapes).

## Demo manifest — MEASURED, `--human-weight 0.2`, seed 20260731, 4.0 s

`wp3_v2_priority/train.jsonl` (depth) x `play_decisions_mixed.jsonl` (breadth). Mixture: **4,808 magezero + 1,202 sampled human = 6,010 records**, realized human share 0.2000 (not capped; 14,659 available).

| metric | magezero | human (mixed) | mixture @ W=0.2 |
|---|---|---|---|
| records (drops) | 4,808 (0) | 14,659 (0) | 6,010 |
| production-shape fraction | 1.0000 | 0.0674 | **0.8128** |
| distinct cards (vocab-validated) | 66 | 1,721 | **1,198** (1,132 new vs magezero) |
| menu size median/mean | 3 / 3.18 | 7 / 7.17 | — |
| gold-action classes | pass 2,026 / cast 1,389 / play 882 / activate 511 | play_land 13,397 / cast 1,158 / pass 76 / activate 24 / other 4 | — |
| best trivial policy | always_land_else_first **0.7633** | always_land_else_first **0.4424** | — |
| unrecognized tokens (distinct / occ) | 10 / 610 | 1,468 / 6,009 | — |

Full-union ceiling (weight -> 1): 1,753 distinct cards. Manifest artifact: `tools/training/data/mixture_manifest_w020.json`.

For the weight decision (numbers, not recommendations):
1. Breadth is cheap: W=0.2 grows the card pool 66 -> 1,198 (18x) while production-shape stays 0.81.
2. The human slice is land-drop heavy: 91.4% of its gold labels are `Play Land:` picks — the 69%-land-drop lesson shape; a land-drop cap knob on the human slice is probably wanted before any training build.
3. The magezero corpus's own reflex ceiling: always_land_else_first 0.7633; always_first_option 0.8032 (verified vs raw records: pick 1 in 3,862/4,808, of which 2,026 have entry 1 = "Pass").
4. Remaining unrecognized tokens are genuine non-cards (Clue/Spirit/Lander tokens) + XMage short-name renderings; DFC faces, `x2`, P/T, `,attacking` suffixes handled (magezero unrecognized occurrences 1,496 -> 610 after fixes).

## How tested

```
pytest tests/test_validate_run_config.py tests/test_build_mixture_manifest.py tests/test_deck_inventory.py -q
56 passed
```
plus the measured demo run above and the live-configs validator run.

### CODE REVIEW

**Comment by josharmour** [comment]
## Adversarial re-verification (fresh worktree at 535a91f, blackwell, venv-train py3.12)

Every claim re-run from a clean `git worktree add` of `origin/wp3-next-run-prep`. Verdict: **MERGEABLE**. No blockers; two should-fix findings below (one resolves an open owner question with a measured answer).

### Reproduced exactly

- **Tests**: the 36 new tests pass; `test_deck_inventory.py` 20 pass (PR body's "56 passed" checks out). Full suite: branch **1489 passed / 4 failed / 29 skipped** vs base 5088e97 **1453 passed / 4 failed / 29 skipped** — the same 4 failures on both (missing `PIL`, FastMCP MagicMock asserts) plus the same `test_draft_guidance_wiring.py` collection error; all pre-existing venv-train environment issues, none introduced by this PR. 1453 + 36 = 1489.
- **Holdout validator, live adversarial**: config with `BGRoots` as opponent → exit 2; `bgroots.dck` variant → exit 2; `HighNoonControl` as primary → exit 2. No-args (tracked configs) → exit 0; `--configs-dir /home/joshu/repos/magezero/configs` → exit 0 over `run.yml` + `run_diverse.yml`, matching the PR transcript. All three holdout `.dck` files exist in `magezero/xmage/decks/`.
- **Demo manifest**: re-ran `build_mixture_manifest.py --human-weight 0.2 --seed 20260731` from the fresh checkout, 3.4 s. Every number matches the self-report and the committed artifact byte-for-byte: 4,808 + 1,202 = 6,010; realized share 0.2000; production-shape 0.8128 (4,885/6,010); distinct cards 66 / 1,721 / 1,150 / 1,198 / 1,132 new / ceiling 1,753; action histograms exact (mz: pass 2,026, cast 1,389, play 882, activate 511; human: play_land 13,397, cast 1,158, pass 76, activate 24, other 4). Deterministic across reruns (manifest identical modulo timestamps).
- **Independent recount, no PR code**: standalone parser over `wp3_v2_priority/train.jsonl` gives pick==1 in **3,862/4,808 (0.8032)**, of which **2,026** have entry 1 = Pass — exact match to concern 2.
- **Corpora inventory**: line counts 4,808 / 14,659 / 988 / 13,671 and all file sizes match; 988/14,659 = 0.0674. Eyeballed raw records of all three shapes (`Legal:` magezero, `Legal:` bridge ground-truth, `Candidate:` 17lands) — classification is correct against the raw text.
- **Card pool**: recomputed from the raw `.dck` files: base union **83**, + diverse five = **235** (+152 new) — exact. All five opponent basenames incl. `Mind(MindvsMight).dck` exist. Live magezero `run_diverse.yml` draft indeed still has `games_per_gen: 200`.
- **Fail-closed rails, exercised live**: missing vocab → exit 2; `--emit` onto existing file → exit 2; `--min-production-share 0.99` → exit 2 with no partial output; manifest-only mode writes no training data. Vocab has 34,633 entries as stated.

### Finding A (answer to the PR's open games_per_gen question — resolve before deploy)

`magezero/src/magezero/runner.py` loops `for opp in run.opponents:` and writes `cfg["training"]["games"] = run.games_per_gen` into each per-opponent `game.yml` (lines 233, 435). **games_per_gen is PER-OPPONENT, not total.** So this config as committed = 10 opponents x 550 x 6 gens = **33,000 games/run**, vs the current run's 5 x 550 x 6 = 16,500. The header's "matches the live run.yml scale" holds per-opponent but doubles total games and wall time. Owner call before the deploy copy: keep 550 (2x budget, equal per-opponent depth) or drop to 275 (equal total budget).

### Finding B (stale doc number)

PROGRESS/PR-body row "unrecognized tokens 10 / 610" for magezero is stale: the committed artifact `mixture_manifest_w020.json` and my rerun both say **18 distinct / 3,947 occurrences**, dominated by ability-text menu entries counted as unresolved activate targets (`{T}: Draw a card...` 1,123, `{1}{U}: Untap` 1,081, `{W/P},` 869), not only Clue/Spirit/Lander tokens. The human row (1,468 / 6,009) and all card-coverage counts are correct as reported; this is a doc-row fix only.

### Minor notes (no action required for merge)

- `scan_corpus` counts a record "usable" (samplable/emittable) once `user`+`response` exist, even if its menu or pick later lands in `drops` — the two overlap. Zero such records in the actual corpora today (`drops: {}` both sources), but a dirtier future corpus could emit menu-less records through `--emit`.
- `--configs-dir` globs `*.yml` only (a `.yaml` run config would be silently skipped in dir mode) and skips `game.yml`/`curriculum.yml` by name; explicit-path invocation and the tests cover the tracked configs.
